File: challs/race_condition.dir/race_condition.py
#!/usr/bin/python3
import sqlite3, sys
import bcrypt, time

## Usage :
# ex : ./race_condition.py register admin password123
# ex : ./race_condition.py login admin password123

## check params
if len(sys.argv) != 4 or not sys.argv[1] or not sys.argv[2] or not sys.argv[3]:
    print("Please send me an 'action' (register or login) with your credentials (login, then password)")
    sys.exit(0)
action = sys.argv[1]
login = sys.argv[2]
passwd = sys.argv[3]
hashed_passwd = str(bcrypt.hashpw(sys.argv[3], bcrypt.gensalt()))
## end check params

try:
	conn = sqlite3.connect('race_condition.db', isolation_level=None)
	cur = conn.cursor()
except:
	print('Error connecting to db.')
	sys.exit(0)


def get_user_id():
	cur.execute("SELECT id, password FROM users WHERE login=?", [login])
	user = cur.fetchone()
	if user:
		if bcrypt.hashpw(passwd, user[1]) == user[1]:
			return user[0]
	return -1


def do_register():
	cur.execute("INSERT INTO users(login, password) VALUES(?, ?)", [login, hashed_passwd])
	user_id = get_user_id()
	cur.execute("INSERT INTO forbidden_ids(user_id) VALUES(?)", [user_id])


def do_login():
	user_id = get_user_id()
	if user_id < 0:
		return "We failed to log you in :/"
	cur.execute("SELECT count(*) FROM forbidden_ids WHERE user_id=?", [user_id])
	if cur.fetchone()[0] > 0:
		return "You are logged in. But sorry you are not allowed to see the secret."
	with open('secret') as s:
		return "You are logged in. And congratz ! Here is the secret : " + s.read()


if action == 'register':
	try:
		do_register()
		print("You are registered !")
	except Exception as e:
		print("An error occurred : " + str(e))
elif action == 'login':
	try:
		print(do_login())
	except Exception as e:
		print("An error occurred : " + str(e))
else:
	print("Error, action param not valid.")

conn.close()

File: challs/race_condition.dir/race_condition.pl
#!/usr/bin/perl -w
use warnings;
use strict;
use Authen::Passphrase::BlowfishCrypt;
use DBI;
use Time::HiRes qw/ time /;

## Usage :
# ex : ./race_condition.pl register admin password123
# ex : ./race_condition.pl login admin password123

my $startTime = time;

## check params
if(@ARGV != 3 || !$ARGV[0] || !$ARGV[1] || !$ARGV[2]){
    print "Please send me an 'action' (register or login) with your credentials (login, then password)\n";
    exit 0;
}
my $action = $ARGV[0];
my $login = $ARGV[1];
my $passwd = $ARGV[2];
## end check params

my $driver = "SQLite";
my $dsn = "DBI:$driver:dbname=/tmp/race_condition.db";
my $dbh = DBI->connect($dsn, undef, undef, { sqlite_see_if_its_a_number => 1, AutoCommit => 1 }) or exit;


sub getUserId{
	my $sth = $dbh->prepare("SELECT id, password FROM users WHERE login=?");
	$sth->bind_param(1, $login);
	$sth->execute;
	$sth->bind_columns(\my($id, $hashedPass));
	my $user = $sth->fetchrow_array;
	if($user){
		my $passphrase = Authen::Passphrase::BlowfishCrypt->from_crypt($hashedPass);
		if($passphrase->match($passwd)) {
			return $id;
		}
	}
	return -1;
}


sub doRegister{
	my $hashedPasswd = Authen::Passphrase::BlowfishCrypt->new( cost=> 8, salt_random => 1, passphrase => $passwd);
	my $sth = $dbh->prepare("INSERT INTO users(login, password) VALUES(?, ?)");
	$sth->bind_param(1, $login);
	$sth->bind_param(2, $hashedPasswd->as_crypt);
	$sth->execute or exit;
	select(undef, undef, undef, 0.5); # simulate more db access / calculus
	my $elapsed = time - $startTime;
	print("It's been " . $elapsed . "s since you started register.\n");
	$sth = $dbh->prepare("INSERT INTO forbidden_ids(user_id) VALUES(?)");
	$sth->execute(getUserId()) or exit;
}


sub doLogin{
	my $userId = getUserId();
	if ($userId < 0){
		return "We failed to log you in :/\n";
	}
	my $sth = $dbh->prepare("SELECT count(*) FROM forbidden_ids WHERE user_id=?");
	my $elapsed = time - $startTime;
	print("It's been " . $elapsed ."s since you started log in.\n");
	$sth->execute($userId) or exit;
	$sth->bind_columns(\my($count));
	$sth->fetchrow_array;
	if($count > 0){
		return "You are logged in. But sorry you are not allowed to see the secret.\n";
	}
	open my $fh, '<', 'secret' or die "error opening secret file";
	my $secret = do { local $/; <$fh> };
	return "You are logged in. And congratz ! Here is the secret : " . $secret . "\n";
}


if ($action eq 'register'){
	doRegister();
	print "You are registered !\n";
}elsif($action eq 'login'){
	print doLogin();
}else{
	print "Error, action param not valid.\n";
}

$dbh->disconnect();

1;
